patroni/consul.py

```diff
@@ -0,0 +1,231 @@
+
+            if results is None:
+                raise NotFound
+
+            nodes = {}
+            for node in results:
+                node['Value'] = node['Value'].decode('utf-8')
+                nodes[os.path.relpath(node['Key'], path)] = node
+
+            # get initialize flag
+            initialize = nodes.get(self._INITIALIZE)
+            initialize = initialize and initialize['Value']
+
+            # get last leader operation
+            last_leader_operation = nodes.get(self._LEADER_OPTIME, None)
+            last_leader_operation = 0 if last_leader_operation is None else int(last_leader_operation['Value'])
+
+            # get list of members
+            members = [self.member(n) for k, n in nodes.items() if k.startswith(self._MEMBERS) and k.count('/') == 1]
+
+            # get leader
+            leader = nodes.get(self._LEADER, None)
+            if leader and leader['Value'] == self._name and self._session != leader.get('Session', 'x'):
+                logger.info('I am leader but not owner of the session. Removing leader node')
+                self._client.kv.delete(self.leader_path, cas=leader['ModifyIndex'])
+                leader = None
+
+            if leader:
+                member = Member(-1, leader['Value'], None, {})
+                member = ([m for m in members if m.name == leader['Value']] or [member])[0]
+                leader = Leader(leader['ModifyIndex'], leader['Session'], member)
+
+            # failover key
+            failover = nodes.get(self._FAILOVER, None)
+            if failover:
+                failover = Failover.from_node(failover['ModifyIndex'], failover['Value'])
+
+            self._cluster = Cluster(initialize, leader, last_leader_operation, members, failover)
+        except NotFound:
+            self._cluster = Cluster(False, None, None, [], None)
+        except:
+            if timeout:
+                raise
+            logger.exception('get_cluster')
+            raise ConsulError('Consul is not responding properly')
+        self._new_cluster = bool(timeout)
+        return self._cluster
+
+    def touch_member(self, connection_string, ttl=None):
+        create_member = self.referesh_session()
+        cluster = self.cluster
+        member_exists = cluster and any(m.name == self._name for m in cluster.members)
+        if create_member and member_exists:
+            self._client.kv.delete(self.member_path)
+        if create_member or not member_exists:
+            try:
+                self._client.kv.put(self.member_path, connection_string, acquire=self._session)
+            except:
+                logger.exception('touch_member')
+        return True
+
+    @catch_consul_errors
+    def attempt_to_acquire_leader(self):
+        ret = self._client.kv.put(self.leader_path, self._name, acquire=self._session)
+        ret or logger.info('Could not take out TTL lock')
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Expression "(ret) or (logger.info('Could not take out TTL lock'))" is assigned to nothing](https://www.codacy.com/app/cyberdemn/patroni/file/2412073583/issues/source?bid=3127594&fileBranchId=3235575#l180)